src/request/plan_builder.rs

```diff
@@ -0,0 +1,178 @@
+/// Used to ensure that a plan has a designated target or targets, a target is
+/// a particular TiKV server.
+pub trait PlanBuilderPhase {}
+pub struct NoTarget;
+impl PlanBuilderPhase for NoTarget {}
+pub struct Targetted;
+impl PlanBuilderPhase for Targetted {}
+
+impl<PdC: PdClient, Req: KvRequest> PlanBuilder<PdC, Dispatch<Req>, NoTarget> {
+    pub fn new(pd_client: Arc<PdC>, request: Req) -> Self {
+        PlanBuilder {
+            pd_client,
+            plan: Dispatch {
+                request,
+                kv_client: None,
+            },
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<PdC: PdClient, P: Plan> PlanBuilder<PdC, P, Targetted> {
+    /// Return the built plan, note that this can only be called once the plan
+    /// has a target.
+    pub fn plan(self) -> P {
+        self.plan
+    }
+}
+
+impl<PdC: PdClient, P: Plan, Ph: PlanBuilderPhase> PlanBuilder<PdC, P, Ph> {
+    /// If there is a lock error, then resolve the lock and retry the request.
+    pub fn resolve_lock(self, backoff: Backoff) -> PlanBuilder<PdC, ResolveLockPlan<P, PdC>, Ph>
+    where
+        P::Result: HasLocks,
+    {
+        PlanBuilder {
+            pd_client: self.pd_client.clone(),
+            plan: ResolveLockPlan {
+                inner: self.plan,
+                backoff,
+                pd_client: self.pd_client,
+            },
+            phantom: PhantomData,
+        }
+    }
+
+    /// If there is a region error, re-shard the request and re-resolve regions, then retry.
+    ///
+    /// Note that this plan must wrap a multi-region plan if the request should be re-sharded.
+    pub fn retry_region(self, backoff: Backoff) -> PlanBuilder<PdC, RetryRegionPlan<P, PdC>, Ph>
+    where
+        P::Result: HasError,
+    {
+        PlanBuilder {
+            pd_client: self.pd_client.clone(),
+            plan: RetryRegionPlan {
+                inner: self.plan,
+                backoff,
+                pd_client: self.pd_client,
+            },
+            phantom: PhantomData,
+        }
+    }
+
+    /// Merge the results of a request. Usually used where a request is sent to multiple regions
+    /// to combine the responses from each region.
+    pub fn merge<In, M: Merge<In>>(self, merge: M) -> PlanBuilder<PdC, MergeResponse<P, In, M>, Ph>
+    where
+        In: Clone + Send + Sync + 'static,
+        P: Plan<Result = Vec<Result<In>>>,
+    {
+        PlanBuilder {
+            pd_client: self.pd_client.clone(),
+            plan: MergeResponse {
+                inner: self.plan,
+                merge,
+                phantom: PhantomData,
+            },
+            phantom: PhantomData,
+        }
+    }
+
+    /// Apply a processing step to a response (usually only needed if the request is sent to a
+    /// single region because post-porcessing can be incorporated in the merge step for multi-region
+    /// requests).
+    pub fn post_process(self) -> PlanBuilder<PdC, ProcessResponse<P, P::Result>, Ph>
```

Comment:
What does this mean?

Reply:
Which part needs clarification? The use case is something like you pass a `RawGetResponse` in and get a `Value` out, i.e., it is discarding some unneeded information and transforming types to something a little bit higher level.

Reply:
I mean, we can do this by default, as there will be no need to call this explicitly,

Reply:
I'm not sure what you mean, we do this by default, but for users who want to go a non-default path, they will need to call it.